kratos/geometries/nurbs_volume_geometry.h

```diff
@@ -0,0 +1,997 @@
+
+        for (IndexType i = 0; i < number_of_spans; i++) {
+            const double t0 = mKnotsW[first_span + i];
+            const double t1 = mKnotsW[first_span + i + 1];
+
+            result[i] = NurbsInterval(t0, t1);
+        }
+
+        return result;
+    }
+    ///@}
+    ///@name Integration Points
+    ///@{
+
+    /** Creates integration points according to the polynomial degrees.
+     * @param return integration points.
+     **/
+    void CreateIntegrationPoints(
+        IntegrationPointsArrayType& rIntegrationPoints) const override
+    {
+        const SizeType points_in_u = PolynomialDegreeU() + 1;
+        const SizeType points_in_v = PolynomialDegreeV() + 1;
+        const SizeType points_in_w = PolynomialDegreeW() + 1;
+
+        CreateIntegrationPoints(
+            rIntegrationPoints, points_in_u, points_in_v, points_in_w);
+    }
+
+    /**
+     * Creates integration points according to the input parameter.
+     * @param return integration points.
+     * @param points in u direction per span.
+     * @param points in v direction per span.
+     * @param points in w direction per span.
+     **/
+    void CreateIntegrationPoints(
+        IntegrationPointsArrayType& rIntegrationPoints,
+        SizeType NumPointsPerSpanU,
+        SizeType NumPointsPerSpanV,
+        SizeType NumPointsPerSpanW) const
+    {
+        auto knot_span_intervals_u = KnotSpanIntervalsU();
+        auto knot_span_intervals_v = KnotSpanIntervalsV();
+        auto knot_span_intervals_w = KnotSpanIntervalsW();
+
+        const SizeType number_of_integration_points =
+            knot_span_intervals_u.size() * knot_span_intervals_v.size() * knot_span_intervals_w.size()
+            * NumPointsPerSpanU * NumPointsPerSpanV * NumPointsPerSpanW;
+
+        if (rIntegrationPoints.size() != number_of_integration_points) {
+            rIntegrationPoints.resize(number_of_integration_points);
+        }
+
+        typename IntegrationPointsArrayType::iterator integration_point_iterator = rIntegrationPoints.begin();
+
+        for (IndexType i = 0; i < knot_span_intervals_u.size(); ++i) {
+            for (IndexType j = 0; j < knot_span_intervals_v.size(); ++j) {
+                for (IndexType k = 0; k < knot_span_intervals_w.size(); ++k) {
+                    IntegrationPointUtilities::IntegrationPoints3D(
+                        integration_point_iterator,
+                        NumPointsPerSpanU, NumPointsPerSpanV, NumPointsPerSpanW,
+                        knot_span_intervals_u[i].GetT0(), knot_span_intervals_u[i].GetT1(),
+                        knot_span_intervals_v[j].GetT0(), knot_span_intervals_v[j].GetT1(),
+                        knot_span_intervals_w[k].GetT0(), knot_span_intervals_w[k].GetT1());
+                }
+            }
+        }
+    }
+
+    Matrix& Jacobian( Matrix& rResult, const CoordinatesArrayType& rCoordinates ) const override
+    {
+        const SizeType working_space_dimension = this->WorkingSpaceDimension();
```

Comment:
No unfortunately we need to override it. The reason is, the shape function values and the number of nodes of the geometry might not coincide, as the shape functions are only computed for non-zero values. It might be discussable if this is good or bad.

In either scenario we can significantly improve the efficiency, here with this function. So it will be always good to have it.

Reply:
Then we also have to add it to the nurbs_surface not only here in the volume?

Reply:
ah now I get your question, yes, it should be there, too. Probably, no one ever needed it... The only reason to have it would be: it would be required to compute the area of the surface.

Reply:
But it is also required to compute the Kinematics inside the element (e.g. inside the
https://github.com/KratosMultiphysics/Kratos/blob/6d7482650222faa9a3b44d6bbaf12c5fd6b986fe/applications/IgaApplication/custom_elements/shell_3p_element.cpp#L210 )